### ISSUE TITLE
Log output is grouped by BuildOperation (task)

### ISSUE THREAD

**Issue by eriwen** [gradle/gradle#1818]
### Expected Behavior
Group related logs together and make it clear which task they belong to (if done within a task). If it cannot be determined where the log came from, just render it live (or consider grouping by phase for Configuration and Initialization). 

Buffer log events until the given `complete` Build Operation event is fired, or the build ends unexpectedly.

### Current Behavior
Logs are rendered asynchronously as they're generated. With `--parallel`, they're often mixed in a way that makes them rather useless.

### Context
This is really part of UX improvements spec'd [here](https://github.com/gradle/gradle/blob/master/design-docs/improved-console.md#milestone-3---logging-improvements)

For a first pass, we should chunk and render output if it crosses one of these thresholds:
 * Number of bytes (2kb?)
 * time (5s?)

However, we ought to reuse the "header" if we're displaying chunked logs from the same task over time. A visual indication is added to the "header" by wrapping it into square brackets. After rendering a chunked log a new line is added for visual separation. For example, a long-running `:test` task may log this after 5 seconds:

```
[:test]
FooSpec FAILED
```

If the same task then logs "BarSpec FAILED", don't render a new `[:test]` header. It should look like this:
```
[:test]
FooSpec FAILED
BarSpec FAILED
```

Nesting of groups should not be allowed initially. _Do not make this behavior specific to tasks._ Instead, provide a declarative way of specifying log "grouping level".

### Test Coverage
* Log messages are flushed upon failure of a build operation
* Log messages are flushed if build is stopped prematurely (via user command or process kill)
* Log messages are only output after operation completes

**Comment by bmuschko** [comment]
Initial spike based on branch `dl-attach-id-to-events`: https://github.com/gradle/gradle/tree/bm-outputevent-grouping

**Comment by bmuschko** [comment]
> Nesting of groups should not be allowed initially. Do not make this behavior specific to tasks. Instead, provide a declarative way of specifying log "grouping level".

I think this requirement needs more clarification. Can you please provide other use cases and examples? What exactly means "declarative" here?

> For a first pass, we should chunk and render output if it crosses one of these thresholds...

Do we really need multiple thresholds here? Especially for the size we'd have to calculate the overall size for every event that is fired. Sounds unnecessarily costly to me.

> Log messages are flushed upon failure of a build operation

I am wondering if we 1) have this information is a consumable format and 2) what type of event that would be. I haven't looked deeper but maybe you have more info on it.

> Log messages are flushed if build is stopped prematurely (via user command or process kill)

Would that always be a `EndOutputEvent`?

> Log messages are only output after operation completes

Probably should say: "Log messages are only output after operation completes or if one of the thresholds is reached - whatever happens first".

**Comment by eriwen** [comment]
@bmuschko I'm afraid I'm probably going to have to make deeper infrastructure changes here. I think it would be best if you picked up some of the other console stories that are orthogonal for now. I can even pick this up from you after if you like.

**Comment by eriwen** [comment]
This has been merged with #1949 and follow-up work scheduled immediately in #1966
